### PR TITLE
feat: move broadcast and spectate to node worker

### PR DESCRIPTION
**Pull request by NikhilNarayana**

### CODE REVIEW

File: src/broadcast/main.ts [hunk 1]
```diff
@@ -1,27 +1,29 @@
...
-  await spectateManager.refreshBroadcastList();
+  const sWorker = await spectateWorker;
+  await sWorker.refreshBroadcastList(authToken);
   return { success: true };
 });
 
 ipc_watchBroadcast.main!.handle(async ({ broadcasterId }) => {
+  const sWorker = await spectateWorker;
   const folderPath = settingsManager.get().settings.spectateSlpPath;
-  spectateManager.watchBroadcast(broadcasterId, folderPath, true);
```
Comment (vinceau) [inline src/broadcast/main.ts:15]: We used to have singleton to be `true` when calling `spectateManager.watchBroadcast()`. This doesn't seem to be the case anymore. Is this intentional?

File: src/broadcast/spectateWorker.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  logSubject.next(msg);
+});
+
+spectateManager.on(SpectateEvent.ERROR, async (errorMsg: string) => {
+  errorSubject.next(errorMsg);
+});
+
+spectateManager.on(SpectateEvent.PLAY_FILE, async (playbackId: string, replayComm: ReplayCommunication) => {
+  spectateDetailsSubject.next({ playbackId, replayComm });
+});
```
Comment (vinceau) [inline src/broadcast/spectateWorker.ts:49]: I'm not really a fan of spectate worker emitting a "play file" event. It shouldn't really need to know about playing things back. I think that should be main's job. The event should probably be called `NEW_GAME` or `NEW_FILE` or `GAME_START` or something which should return an Observable string (the SLP filename). then it should be main's job to generate the playback ID and replay communication.

File: src/broadcast/types.ts [hunk 3]
```diff
@@ -14,3 +16,46 @@ export interface StartBroadcastConfig {
   authToken: string;
   name?: string;
 }
+
+export enum BroadcastEvent {
+  slippiStatusChange = "slippiStatusChange",
+  dolphinStatusChange = "dolphinStatusChange",
+  error = "broadcastError",
+  LOG = "log",
```
Comment (vinceau) [inline src/broadcast/types.ts:24]: The log event doesn't match the naming convention of the other events. We should probably coerce the others to be in upper snake case or leave it for another PR.

File: src/broadcast/workerInterfaces.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+const broadcastLog = electronLog.scope("broadcastManager");
+const spectateLog = electronLog.scope("spectateManager");
+
+export const broadcastWorker: Promise<Thread & BroadcastWorkerMethods> = new Promise((resolve, reject) => {
+  log.debug("broadcast: Spawning worker");
+
+  spawn<BroadcastWorkerSpec>(new Worker("./broadcastWorker"), { timeout: 30000 })
+    .then((worker) => {
+      worker.getDolphinStatusObservable().subscribe(({ status }) => {
+        log.info(`got dolphin status: ${status}`);
```
Comment (vinceau) [inline src/broadcast/workerInterfaces.ts:26]: Should we remove these logs? I think they were useful for debugging but they're probably not useful now.

**Review by vinceau — CHANGES_REQUESTED** [review]

File: src/broadcast/workerInterfaces.ts (hunk 4 above)
Comment (NikhilNarayana) [reply src/broadcast/workerInterfaces.ts:26]: yeah we can clean those up

File: src/broadcast/types.ts (hunk 3 above)
Comment (NikhilNarayana) [reply src/broadcast/types.ts:24]: whoops, didn't notice the difference when I copied it from SpectateEvents.

File: src/broadcast/main.ts (hunk 1 above)
Comment (NikhilNarayana) [reply src/broadcast/main.ts:15]:
```ts
  public watchBroadcast(broadcastId: string, targetPath: string, singleton?: true)
  ```
 the function signature forces true so I skipped it when i was rewriting. Do we want to put true anyways?

File: src/broadcast/spectateWorker.ts (hunk 2 above)
Comment (NikhilNarayana) [reply src/broadcast/spectateWorker.ts:49]: Sure

File: src/broadcast/main.ts (hunk 1 above)
Comment (vinceau) [reply src/broadcast/main.ts:15]: I didn't realise it was true by default. It's fine to leave it as undefined.

File: src/broadcast/broadcastManager.ts [hunk 5]
```diff
@@ -83,20 +82,20 @@ export class BroadcastManager {
       try {
         await this._connectToDolphin(config.ip, config.port);
       } catch (err) {
-        log.warn(err);
+        this.emit(BroadcastEvent.LOG, err);
```
Comment (vinceau) [inline src/broadcast/broadcastManager.ts:85]: Error is an object. I'm pretty sure subscribers will be expecting a string. So we should make sure we're only return a string.

File: src/broadcast/broadcastManager.ts [hunk 6]
```diff
@@ -70,7 +69,7 @@ export class BroadcastManager {
     });
     this.dolphinConnection.on(ConnectionEvent.ERROR, (err) => {
       // Log the error messages we get from Dolphin
-      log.error("[Broadcast] Dolphin connection error\n", err);
+      this.emit(BroadcastEvent.LOG, "[Broadcast] Dolphin connection error\n", err);
```
Comment (vinceau) [inline src/broadcast/broadcastManager.ts:72]: Again over here, this event should only emit a single string.  We're returning 2 things here, a string, and an error object.

File: src/broadcast/types.ts [hunk 7]
```diff
@@ -14,3 +16,46 @@ export interface StartBroadcastConfig {
   authToken: string;
   name?: string;
 }
+
+export enum BroadcastEvent {
+  SLIPPI_STATUS_CHANGE = "slippiStatusChange",
+  DOLPHIN_STATUS_CHANGE = "dolphinStatusChange",
+  ERROR = "broadcastError",
+  LOG = "log",
```
Comment (vinceau) [inline src/broadcast/types.ts:24]:
```suggestion
  SLIPPI_STATUS_CHANGE = "SLIPPI_STATUS_CHANGE",
  DOLPHIN_STATUS_CHANGE = "DOLPHIN_STATUS_CHANGE",
  ERROR = "ERROR",
  LOG = "LOG",
```

While we're refactoring these, can we make it like this?

File: src/broadcast/types.ts [hunk 8]
```diff
@@ -14,3 +16,46 @@ export interface StartBroadcastConfig {
...
+}
+
+export enum SpectateEvent {
+  ERROR = "error",
+  BROADCAST_LIST_UPDATE = "broadcastListUpdate",
+  NEW_FILE = "newFile",
+  LOG = "log",
```
Comment (vinceau) [inline src/broadcast/types.ts:31]:
```suggestion
  ERROR = "ERROR",
  BROADCAST_LIST_UPDATE = "BROADCAST_LIST_UPDATE",
  NEW_FILE = "NEW_FILE",
  LOG = "LOG",
```

**Review by vinceau — CHANGES_REQUESTED** [review]
Just some minor fixes.

File: src/broadcast/broadcastManager.ts (hunk 5 above)
Comment (vinceau) [reply src/broadcast/broadcastManager.ts:85]: We can probably log `err.message || JSON.stringify(err)`.

File: src/broadcast/spectateManager.ts [hunk 9]
```diff
@@ -131,23 +98,25 @@ export class SpectateManager extends EventEmitter {
     await new Promise<void>((resolve, reject) => {
       const socket = new WebSocketClient();
 
-      socket.on("connectFailed", (error) => {
-        log.error(`[Spectate] WS connection failed\n`, error);
-        this.emit(SpectateManagerEvent.ERROR, error);
+      socket.on("connectFailed", (err) => {
+        const errMsg = err.message || JSON.stringify(err);
+        this.emit(SpectateEvent.LOG, `WS connection failed\n${errMsg}`);
+        this.emit(SpectateEvent.ERROR, errMsg);
```
Comment (vinceau) [inline src/broadcast/spectateManager.ts:104]: Can we use the `log` event for logs and the `error` event for errors? It seems like in a lot of places we're doing the double emit to log and error the same message.

**Review by vinceau — APPROVED** [review]
LGTM
